### DISCUSSION TITLE
Use an existing plugin proxy to retrieve data

### DISCUSSION THREAD

**Discussion by BarryDrez** [eclipse-theia/theia#12552] [General]
Is there any way that I can make a call to retrieve data from a plugin?  For example, I would like to be able use something like the proxy of theia/packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx to call:

`proxy.$getChildren("javaProjectExplorer", "'/iproject/iReferenced Libraries'");`

I know that normally the Java Projects must be visible before it calls:

`proxy.$setVisible("javaProjectExplorer", true);`

Can I call this if the view is not visible so I can use the other proxy functions?

**Comment by msujew** [comment]
@BarryDrez Communication between the Theia frontend and vscode extensions can be accomplished using commands. Any command contributed by a vscode extension can be called from the frontend and vice versa.

**Comment by BarryDrez** [comment]
Hi @msujew, Thanks for the reply.  In this case, there are no actions being invoked by the frontend.  there are only calls going through the proxy.

**Reply by msujew** [reply]
Do you mind explaining a bit what you're trying to achieve here? Which component is trying to retrieve data from a plugin?

**Comment by BarryDrez** [comment]
@msujew, I am trying to retrieve the data for Referenced Libraries within the JAVA PROJECTS view:

![image](https://github.com/eclipse-theia/theia/assets/24534027/b2456881-5acc-4bf2-9348-f142634da08e)

I have to be able to drill down just as the tree does to do some analysis at each node level.  Any tips on how to do this would be greatly appreciated.
Thank you!

**Reply by BarryDrez** [reply]
Just FYI: I solved this by creating my own server RPC for plugin host "main" with a channel.  E.g.:
`const rpc = new RPCProtocolImpl(channel);`

Then I got the proxy using:
`rpc.getProxy(MAIN_RPC_CONTEXT.TREE_VIEWS_EXT);`

Probably not the best way to do this since it duplicates Theia code in plugin-ext, but it solves my problem.
